xenomint/state.go

```diff
@@ -0,0 +1,617 @@
+				// Add to failed pool list
+				s.pool.setFailed(req)
+				s.rollbackTo(savepoint)
+				return
+			}
+
+			curAffectedRows, _ = res.RowsAffected()
+			lastInsertID, _ = res.LastInsertId()
+			totalAffectedRows += curAffectedRows
+		}
+		s.setSavepoint()
+		s.pool.enqueue(savepoint, query)
+		return
+	}(); err != nil {
+		return
+	}
+	// Build query response
+	ref = query
+	resp = &types.Response{
+		Header: types.SignedResponseHeader{
+			ResponseHeader: types.ResponseHeader{
+				Request:      req.Header,
+				NodeID:       s.nodeID,
+				Timestamp:    s.getLocalTime(),
+				RowCount:     0,
+				LogOffset:    savepoint,
+				AffectedRows: totalAffectedRows,
+				LastInsertID: lastInsertID,
+			},
+		},
+	}
+	return
+}
+
+func (s *State) replay(req *types.Request, resp *types.Response) (err error) {
+	var (
+		ierr      error
+		savepoint uint64
+		query     = &QueryTracker{Req: req, Resp: resp}
+	)
+	s.Lock()
+	defer s.Unlock()
+	savepoint = s.getID()
+	if resp.Header.ResponseHeader.LogOffset != savepoint {
+		err = errors.Wrapf(
+			ErrQueryConflict,
+			"local id %d vs replaying id %d", savepoint, resp.Header.ResponseHeader.LogOffset,
+		)
+		return
+	}
+	for i, v := range req.Payload.Queries {
+		if _, ierr = s.writeSingle(&v); ierr != nil {
+			err = errors.Wrapf(ierr, "execute at #%d failed", i)
+			s.rollbackTo(savepoint)
+			return
+		}
+	}
+	s.setSavepoint()
+	s.pool.enqueue(savepoint, query)
+	return
+}
+
+// ReplayBlock replays the queries from block. It also checks and skips some preceding pooled
+// queries.
+func (s *State) ReplayBlock(block *types.Block) (err error) {
+	var (
+		ierr   error
+		lastsp uint64 // Last savepoint
+	)
+	s.Lock()
+	defer s.Unlock()
+	for i, q := range block.QueryTxs {
+		var query = &QueryTracker{Req: q.Request, Resp: &types.Response{Header: *q.Response}}
+		lastsp = s.getID()
+		if q.Response.ResponseHeader.LogOffset > lastsp {
+			err = ErrMissingParent
+			return
+		}
+		// Match and skip already pooled query
+		if q.Response.ResponseHeader.LogOffset < lastsp {
+			if !s.pool.match(q.Response.ResponseHeader.LogOffset, q.Request) {
+				err = ErrQueryConflict
+				return
+			}
+			continue
+		}
+		// Replay query
+		for j, v := range q.Request.Payload.Queries {
+			if q.Request.Header.QueryType == types.ReadQuery {
+				continue
+			}
+			if q.Request.Header.QueryType != types.WriteQuery {
+				err = errors.Wrapf(ErrInvalidRequest, "replay block at %d:%d", i, j)
+				s.rollbackTo(lastsp)
+				return
+			}
+			if _, ierr = s.writeSingle(&v); ierr != nil {
+				err = errors.Wrapf(ierr, "execute at %d:%d failed", i, j)
+				s.rollbackTo(lastsp)
+				return
+			}
+		}
+		s.setSavepoint()
+		s.pool.enqueue(lastsp, query)
+	}
+	// Remove duplicate failed queries from local pool
+	for _, r := range block.FailedReqs {
+		s.pool.removeFailed(r)
+	}
+	// Check if the current transaction is ok to commit
+	if s.pool.matchLast(lastsp) {
+		if err = s.uncCommit(); err != nil {
+			// FATAL ERROR
```

Comment:
It actually needs some further processing here, or it may run into an unrecoverable state. And usually, a Tx without context doesn't produce error on committing. Maybe leave a TODO here for later optimization.